### ISSUE TITLE
VProc Deadlock

### ISSUE THREAD

**Issue by ml9951** [ManticoreProject/manticore#1]
The work stealing scheduler can still deadlock despite the previously proposed "fix" of playing with the number of idle VProcs.  

Currently, we set `nIdle` in the workgroup to `-1` so the work group cannot shutdown before the first task is put on the deque.  Once the first task is placed on the deque, we atomically increment `nIdle`.  

I believe the problem occurs in the following sequence of events

1. The work stealing instance on the main VProc goes to sleep
2. Place the main task on the main VProc's deque
3. Atomically increment `nIdle`
4. Remaining threads go to sleep
5. Work group shuts down without running the first task.

**Comment by ml9951** [comment]
Currently, to start the work group, we have:

```
    (* enter the dynamic scope of a work group *)
      define inline @work-group-begin (group : work_group / exh : exh) : () =
	  do @migrate-to-top-level-sched (/ exh)
	  do @push-work-group (group / exh)
	(* migrate the current thread to the given work group *)
	  cont k (x : unit) = return ()
	  let thd : thread = @new-thread (k / exh)
	  do @spawn-thread (thd / exh)
          let nIdle : ![int] = #WORK_GROUP_N_IDLE_OFF(group)
          let _ : int = I32FetchAndAdd(&0(nIdle), 1)
	  let _ : unit = SchedulerAction.@stop ()
	  return ()
	;
```

If tasks are guaranteed to be stolen only once, then this could work:

```
    (* enter the dynamic scope of a work group *)
      define inline @work-group-begin (group : work_group / exh : exh) : () =
	  do @migrate-to-top-level-sched (/ exh)
	  do @push-work-group (group / exh)
	(* migrate the current thread to the given work group *)
	  cont k (x : unit) = 
              let nIdle : ![int] = #WORK_GROUP_N_IDLE_OFF(group)
              let _ : int = I32FetchAndAdd(&0(nIdle), 1)
              return ()
	  let thd : thread = @new-thread (k / exh)
	  do @spawn-thread (thd / exh)		
	  let _ : unit = SchedulerAction.@stop ()
	  return ()
	;
```

**Comment by ml9951** [comment]
Otherwise, I think this should work:

```
   (* enter the dynamic scope of a work group *)
      define inline @work-group-begin (group : work_group / exh : exh) : () =
	  do @migrate-to-top-level-sched (/ exh)
	  do @push-work-group (group / exh)
	  let bumped : ![bool] = alloc(false)
	  let bumped : ![bool] = promote(bumped)
	(* migrate the current thread to the given work group *)
	  cont k (x : unit) = 
	      if BCAS(&0(bumped), true, false)
	      then
		  do ccall M_Print("incrementing\n")
		  let nIdle : ![int] = #WORK_GROUP_N_IDLE_OFF(group)
                  let _ : int = I32FetchAndAdd(&0(nIdle), 1)
                  return()
	      else return ()
	  let thd : thread = @new-thread (k / exh)
	  do @spawn-thread (thd / exh)		
	  let _ : unit = SchedulerAction.@stop ()
	  return ()
	;
```

**Comment by kavon** [comment]
If the 2nd bugfix on trunk didn't fix it, can you revert that commit (5bcf4fc8792) on trunk and then make a new branch off of trunk to work on a fix?

**Comment by ml9951** [comment]
Fixed in 26e704dee9032ed897dfe9388595d4df1e290136
